### PR TITLE
APR-2916 confirm organisation trustees refresh

### PR DESCRIPTION
**Pull request by MarkFCain**

### CODE REVIEW

**Review by raviudari01 — APPROVED** [review]

**Comment by raviudari01** [comment]
Please check code smell

File: src/SFA.DAS.ApplyService.Web/Infrastructure/Services/RefreshTrusteesService.cs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            {
+                throw new InvalidOperationException($"RefreshTrusteesService for Application {applicationId} update trustees failed: [{ex.Message}]");
+
+            }
+
+            var applicationDetails = new Domain.Roatp.ApplicationDetails { CharitySummary = Mapper.Map<CharityCommissionSummary>(charityDetails) };
+            var trusteesAnswers = RoatpPreambleQuestionBuilder.CreateCharityCommissionWhosInControlQuestions(applicationDetails);
+
+            _logger.LogInformation($"RefreshTrusteesService: resetting page answers for charities for page 80, 86 and section 3.4, applicationId {applicationId}");
+            var resetSection1_3_80 = _qnaApiClient.ResetPageAnswersBySequenceAndSectionNumber(applicationId, RoatpWorkflowSequenceIds.YourOrganisation, RoatpWorkflowSectionIds.YourOrganisation.WhosInControl, RoatpWorkflowPageIds.WhosInControl.CharityCommissionTrustees);
```
Comment (DevenKShah) [inline src/SFA.DAS.ApplyService.Web/Infrastructure/Services/RefreshTrusteesService.cs:91]: Why no exception handling during reset requests?

File: src/SFA.DAS.ApplyService.Web.UnitTests/Services/RefreshTrusteeServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+            _qnaApiClient.Verify(x => x.UpdatePageAnswers(_applicationId, RoatpWorkflowSequenceIds.YourOrganisation, RoatpWorkflowSectionIds.YourOrganisation.WhosInControl, RoatpWorkflowPageIds.WhosInControl.CharityCommissionTrustees, It.IsAny<List<Answer>>()), Times.Once);
+            _qnaApiClient.Verify(x => x.ResetPageAnswersBySequenceAndSectionNumber(_applicationId, RoatpWorkflowSequenceIds.YourOrganisation, RoatpWorkflowSectionIds.YourOrganisation.WhosInControl, RoatpWorkflowPageIds.WhosInControl.CharityCommissionTrustees), Times.Once);
+            _qnaApiClient.Verify(x => x.ResetPageAnswersBySequenceAndSectionNumber(_applicationId, RoatpWorkflowSequenceIds.YourOrganisation, RoatpWorkflowSectionIds.YourOrganisation.WhosInControl, RoatpWorkflowPageIds.WhosInControl.CharityCommissionTrusteesDob), Times.Once);
+            _qnaApiClient.Verify(x => x.ResetPageAnswersBySection(_applicationId, RoatpWorkflowSequenceIds.CriminalComplianceChecks, RoatpWorkflowSectionIds.CriminalComplianceChecks.CheckOnWhosInControl), Times.Once);
+        }
+
+        [Test]
+        public void refresh_trustees_and_update_trustees_fails_exception_thrown()
```
Comment (DevenKShah) [inline src/SFA.DAS.ApplyService.Web.UnitTests/Services/RefreshTrusteeServiceTests.cs:197]: Missing test for when you are raising error on update of organisation.

File: src/SFA.DAS.ApplyService.Web.UnitTests/Controllers/RoatpWhosInControlRefreshControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            {
+                ControllerContext = new ControllerContext()
+                {
+                    HttpContext = new DefaultHttpContext() { User = user },
+                }
+            };
+        }
+
+        [Test]
+        public async Task refresh_trustees_and_redirect_if_details_not_available()
```
Comment (DevenKShah) [inline src/SFA.DAS.ApplyService.Web.UnitTests/Controllers/RoatpWhosInControlRefreshControllerTests.cs:53]: Need to follow naming convention here.

**Review by DevenKShah — COMMENTED** [review]
Few more observations.

File: src/SFA.DAS.ApplyService.Data/OrganisationRepository.cs [hunk 4]
```diff
@@ -66,8 +66,8 @@ public async Task<Organisation> GetOrganisationByApplicationId(Guid applicationI
                 var sql =
                     @"SELECT Organisations.*
                     FROM Organisations 
-                    INNER JOIN Applications ON Applications.ApplyingOrganisationId = Organisations.Id
-                    WHERE Applications.Id = @applicationId";
+                    INNER JOIN Apply ON Apply.OrganisationId = Organisations.Id
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApplyService.Data/OrganisationRepository.cs:69]: I know, right? I assume it was lifted from qna or something,....

File: src/SFA.DAS.ApplyService.Application/Organisations/GetOrganisation/GetOrganisationByApplicationIdRequest.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+namespace SFA.DAS.ApplyService.Application.Organisations.GetOrganisation
+{
+    public class GetOrganisationByApplicationIdRequest : IRequest<Organisation>
+    {
+        public GetOrganisationByApplicationIdRequest(Guid applicationId)
+        {
+            ApplicationId = applicationId;
+        }
+
+        public Guid ApplicationId { get; set; }
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApplyService.Application/Organisations/GetOrganisation/GetOrganisationByApplicationIdRequest.cs:14]: Can't be private as is needed, but the set isn't needed, so removed.

File: src/SFA.DAS.ApplyService.Web/Infrastructure/Services/RefreshTrusteesService.cs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+            var charityNumber = organisation?.OrganisationDetails?.CharityNumber;
+
+            if (ukprn == null || charityNumber == null || !int.TryParse(charityNumber, out var charityNumberValue))
+            {
+                _logger.LogInformation($"RefreshTrusteesService: Refresh failure for applicationId {applicationId}, ukprn: [{ukprn}], Charity number: [{charityNumber}]");
+                
+                return new RefreshTrusteesResult { CharityDetailsNotFound = true, CharityNumber = charityNumber };
+            }
+
+            var currentStatusesForUkrpn = await _applicationApiClient.GetExistingApplicationStatus(ukprn);
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApplyService.Web/Infrastructure/Services/RefreshTrusteesService.cs:49]: I've DM'd you about this separately

File: src/SFA.DAS.ApplyService.Web/Infrastructure/Services/RefreshTrusteesService.cs (hunk 1 above)
Comment (MarkFCain) [reply src/SFA.DAS.ApplyService.Web/Infrastructure/Services/RefreshTrusteesService.cs:91]: I modelled it on the work done in company, but looking at it I realise that has a big try-catch.  Handling now added.

File: src/SFA.DAS.ApplyService.Application/Organisations/GetOrganisation/GetOrganisationByApplicationIdRequest.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (DevenKShah) [reply src/SFA.DAS.ApplyService.Application/Organisations/GetOrganisation/GetOrganisationByApplicationIdRequest.cs:14]: 😆

File: src/SFA.DAS.ApplyService.Web.UnitTests/Services/RefreshTrusteeServiceTests.cs (hunk 2 above)
Comment (MarkFCain) [reply src/SFA.DAS.ApplyService.Web.UnitTests/Services/RefreshTrusteeServiceTests.cs:197]: The update trustee call updates the organisation.

File: src/SFA.DAS.ApplyService.InternalApi/Controllers/OrganisationController.cs [hunk 7]
```diff
@@ -122,6 +136,19 @@ public async Task<ActionResult<Organisation>> GetOrganisationById(Guid organisat
             return Ok(org);
         }
 
+        [HttpGet("applicationId/{applicationId}")]
+        public async Task<ActionResult<Organisation>> GetOrganisationByApplicationId(Guid applicationId)
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApplyService.InternalApi/Controllers/OrganisationController.cs:140]: Coverage added

File: src/SFA.DAS.ApplyService.Web.UnitTests/Controllers/RoatpWhosInControlRefreshControllerTests.cs (hunk 3 above)
Comment (MarkFCain) [reply src/SFA.DAS.ApplyService.Web.UnitTests/Controllers/RoatpWhosInControlRefreshControllerTests.cs:53]: updated

File: src/SFA.DAS.ApplyService.Application.UnitTests/Handlers/UpdateOrganisationTrusteesHandlerTests/UpdateOrganisationTrusteesHandlerTests.cs [hunk 8]
```diff
@@ -0,0 +1,78 @@
+﻿
```
Comment (DevenKShah) [inline src/SFA.DAS.ApplyService.Application.UnitTests/Handlers/UpdateOrganisationTrusteesHandlerTests/UpdateOrganisationTrusteesHandlerTests.cs:1]: ⚔️

File: src/SFA.DAS.ApplyService.Application/Organisations/GetOrganisation/GetOrganisationByApplicationIdRequest.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+{
+    public class GetOrganisationByApplicationIdRequest : IRequest<Organisation>
+    {
+        public Guid ApplicationId { get;  }
+        public GetOrganisationByApplicationIdRequest(Guid applicationId)
+        {
+            ApplicationId = applicationId;
+        }
+
+
```
Comment (DevenKShah) [inline src/SFA.DAS.ApplyService.Application/Organisations/GetOrganisation/GetOrganisationByApplicationIdRequest.cs:15]: ⚔️

File: src/SFA.DAS.ApplyService.InternalApi.UnitTests/Controllers/OrganisationControllerTests/GetOrganisationByApplicationIdTests.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+using NUnit.Framework;
+using SFA.DAS.ApplyService.Application.Organisations.GetOrganisation;
+using SFA.DAS.ApplyService.Application.Organisations.UpdateOrganisation;
+using SFA.DAS.ApplyService.Domain.Entities;
+using SFA.DAS.ApplyService.InternalApi.Controllers;
+
+namespace SFA.DAS.ApplyService.InternalApi.UnitTests.Controllers.OrganisationControllerTests
+{
+
+
```
Comment (DevenKShah) [inline src/SFA.DAS.ApplyService.InternalApi.UnitTests/Controllers/OrganisationControllerTests/GetOrganisationByApplicationIdTests.cs:21]: ⚔️

File: src/SFA.DAS.ApplyService.Web.UnitTests/Services/RefreshTrusteeServiceTests.cs [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+            _qnaApiClient.Verify(x => x.UpdatePageAnswers(_applicationId, RoatpWorkflowSequenceIds.YourOrganisation, RoatpWorkflowSectionIds.YourOrganisation.WhosInControl, RoatpWorkflowPageIds.WhosInControl.CharityCommissionTrustees, It.IsAny<List<Answer>>()), Times.Once);
+            _qnaApiClient.Verify(x => x.ResetPageAnswersBySequenceAndSectionNumber(_applicationId, RoatpWorkflowSequenceIds.YourOrganisation, RoatpWorkflowSectionIds.YourOrganisation.WhosInControl, RoatpWorkflowPageIds.WhosInControl.CharityCommissionTrustees), Times.Once);
+            _qnaApiClient.Verify(x => x.ResetPageAnswersBySequenceAndSectionNumber(_applicationId, RoatpWorkflowSequenceIds.YourOrganisation, RoatpWorkflowSectionIds.YourOrganisation.WhosInControl, RoatpWorkflowPageIds.WhosInControl.CharityCommissionTrusteesDob), Times.Once);
+            _qnaApiClient.Verify(x => x.ResetPageAnswersBySection(_applicationId, RoatpWorkflowSequenceIds.CriminalComplianceChecks, RoatpWorkflowSectionIds.CriminalComplianceChecks.CheckOnWhosInControl), Times.Once);
+        }
+
+        [Test]
+        public void RefreshTrustees_UpdateTrusteesReturnsFalse_ExceptionThrown()
+
```
Comment (DevenKShah) [inline src/SFA.DAS.ApplyService.Web.UnitTests/Services/RefreshTrusteeServiceTests.cs:193]: ⚔️

File: src/SFA.DAS.ApplyService.Web.UnitTests/Services/RefreshTrusteeServiceTests.cs [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+            _service = new RefreshTrusteesService(_qnaApiClient.Object,
+                _outerApiClient.Object,
+                _organisationApiClient.Object,
+                _applicationApiClient.Object,
+                _logger.Object
+            );
+        }
+
+
```
Comment (DevenKShah) [inline src/SFA.DAS.ApplyService.Web.UnitTests/Services/RefreshTrusteeServiceTests.cs:58]: ⚔️

File: src/SFA.DAS.ApplyService.Web.UnitTests/Services/RefreshTrusteeServiceTests.cs [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+using SFA.DAS.ApplyService.Domain.Roatp;
+using SFA.DAS.ApplyService.Infrastructure.ApiClients;
+using SFA.DAS.ApplyService.InternalApi.Types.CharityCommission;
+using SFA.DAS.ApplyService.Web.AutoMapper;
+using SFA.DAS.ApplyService.Web.Infrastructure;
+using SFA.DAS.ApplyService.Web.Infrastructure.Services;
+
+namespace SFA.DAS.ApplyService.Web.UnitTests.Services
+{
+
```
Comment (DevenKShah) [inline src/SFA.DAS.ApplyService.Web.UnitTests/Services/RefreshTrusteeServiceTests.cs:22]: ⚔️

**Review by DevenKShah — APPROVED** [review]
Looks fab. I have marked few places where I found extra empty lines, if you want to tidy up ;-)

**Review by DevenKShah — DISMISSED** [review]

**Review by DevenKShah — APPROVED** [review]

**Review by DevenKShah — APPROVED** [review]

**Pull request merged**